apps/www/content/glossary/api-rate-limiting.mdx

```diff
@@ -0,0 +1,150 @@
+
+## Understanding API Rate Limiting Concepts
+
+API rate limiting involves restricting the number of API calls a user or service can make within a specified time period. Key concepts include:
+
+- **Rate Limit**: The maximum number of requests a client can make in a defined time period.
+- **Throttle**: Temporarily limiting the access of users who exceed the rate limit.
+- **Quota**: The total number of allowed requests in a longer time period (e.g., a day or a month).
+- **Burst Limits**: Allows for short bursts of calls, often used to handle usage spikes smoothly.
+
+## Best Practices for API Rate Limiting
+
+Implementing effective **API rate limiting best practices** is crucial for maintaining API performance and user satisfaction. Here are some recommended strategies:
+
+1. **Transparent Policies**: Clearly communicate rate limits to API consumers to avoid confusion and frustration.
+2. **Dynamic Limits**: Adjust rate limits based on server load and user behavior to optimize performance.
+3. **High Availability**: Implement failover strategies to ensure limit enforcement when primary systems fail.
+4. **Graceful Degradation**: Instead of hard stopping, provide users with limited functionality or degraded service when they hit limits.
+5. **Feedback Mechanisms**: Use HTTP headers to provide real-time usage data back to the user.
+
+## Implementing API Rate Limiting: Examples
+
+Here’s a practical example of implementing **API rate limiting** in Python using Flask and Redis:
+
+```python
+# Python example using Flask and Redis to implement rate limiting
+from flask import Flask, request, jsonify
+import redis
+from functools import wraps
+
+app = Flask(__name__)
+redis_client = redis.StrictRedis(host='localhost', port=6379, db=0)
+
+def rate_limiter(max_requests, time_period):
+    def decorator(f):
+        @wraps(f)
+        def decorated_function(*args, **kwargs):
+            key = f"{request.remote_addr}:{request.endpoint}"
+            try:
+                requests = int(redis_client.get(key) or 0)
+            except redis.RedisError:
+                return jsonify({"error": "Redis server is not responding"}), 503
+
+            if requests >= max_requests:
+                return jsonify({"error": "Rate limit exceeded"}), 429
+
+            redis_client.incr(key, 1)
+            redis_client.expire(key, time_period)
+            return f(*args, **kwargs)
+        return decorated_function
+    return decorator
+
+@app.route('/api/resource')
+@rate_limiter(max_requests=100, time_period=60)
+def my_api_endpoint():
+    return jsonify({"success": "Access granted"})
+
+if __name__ == "__main__":
+    app.run()
+```
+
+## GitHub Resources for API Rate Limiting
```

Comment:
Do we expect the user to search for these?
can we use links instead?
@p6l-richard

Reply:
Noting:
- exclude GitHub (or any particular) Resources from content generation
- rename the "recommended reading" section from takeaways into "recommended resources" and have it include github link(s) if fitting